### ISSUE TITLE
Code Coverage should not be checking the unit test projects themselves for coverage. 

### ISSUE THREAD

**Issue by StingyJack** [microsoft/vstest#1548]
Running the code coverage option from the test explorer will always display the coverage statistics for unit test assemblies, and this skews the overall results. 

I'm aware that this can be excluded by using a runsettings/testsettings file, but that has become tedious after a hundred or so projects, especially since VS cant associate and remember the test settings file with the solution. The way it requires file naming is also easy to mess up.

There is a request open to permit the [ExcludeFromCodeCoverage] attribute to an assembly, but that is targeted to net core and not net framework. 

Can unit test projects just be removed from the coverage stats by default, and still let those who want this opt in?

**Comment by smadala** [comment]
Agreed with @StingyJack , To get proper coverage for product code only, Currently we have to pass runsettings. I run into same experience when try to get code coverage for vstest repo.

**Comment by gioce90** [comment]
This issue is open by 2018. I'm using collect:"Code coverage" (the new native ms dotnet coverage tool) and I create cobertura or . coverage results as well, but is painful using . runsettings configuration file for exclude tests and 3rd party libraries.

**Comment by devlie** [comment]
`.runsettings` workaround is not ideal because its default is to include all binaries including nuget dependencies, which means we have to specify both `<Include>` to include our namespace and `<Exclude>` to exclude out tests.

We tried specifying assembly level `ExcludeFromCodeCoverage`, but that didn't work. It doesn't look like it's respected by VS/dotnet test.

**Comment by IvanAlekseev** [comment]
Is there any update on this one? I am trying to migrate from coverlet and using runsettings file doesn't feel good. It would be awesome to have some clever defaults. At least stop trying to cover all nuget dependencies

**Comment by devlie** [comment]
> Is there any update on this one? I am trying to migrate from coverlet and using runsettings file doesn't feel good. It would be awesome to have some clever defaults. At least stop trying to cover all nuget dependencies

@IvanAlekseev , please upvote https://github.com/microsoft/testfx/issues/1060 and https://github.com/microsoft/vstest/issues/1050 so we can get some traction :)

**Comment by Evangelink** [comment]
Cc @jakubch1 @MarcoRossignoli

**Comment by MarcoRossignoli** [comment]
It's available today going to close https://github.com/microsoft/vstest/issues/3669#issuecomment-1143484050

**Comment by bramborman** [comment]
@MarcoRossignoli this should IMO be reopened as it doesn't fix the original issue - to automatically exclude test projects from the coverage - that is not implemented. We still need to manually add the `ExcludeFromCodeCoverageAttribute` assembly-wide, but that is not possible on older target frameworks.

**Comment by MarcoRossignoli** [comment]
@bramborman can you show your config? by default coverlet excludes the test project(if it's not referenced in some way to another one and so it's taken as dependency and instrumented).
Are you on the last version?

**Comment by jakubch1** [comment]
Unfortunately, Microsoft.CodeCoverage package covers by default all assemblies with pdb existing. You can exclude using `ExcludeFromCodeCoverageAttribute` or runsettings. If you don't want to generate runsettings you can also use this way:
```
dotnet test --collect "Code Coverage" -- DataCollectionRunSettings.DataCollectors.DataCollector.Configuration.CodeCoverage.ModulePaths.Exclude.ModulePath=".*tests.dll"
```

**Comment by bramborman** [comment]
Yeah I know this is possible, but think it should be excluded by default, as it doesn't really make sense in almost all of the possible cases I believe. The attribute is not available to be applied assembly-wide in older frameworks, so that's not a solution, and applying it class-by-class is tedious and error-prone... Runsettings are a way, but once again, I believe almost everybody using this doesn't want to have coverage counted for the test projects, so it's just an annoying repetitive stuff to add for every test run everywhere... It should be configurable, but rather as an opt-in, than opt-out IMHO...

**Comment by jakubch1** [comment]
The issue is that many teams are collecting coverage also for test projects. Changing default behavior could break many solutions.

**Comment by StingyJack** [comment]
@jakubch1 - are you being serious or sarcastic?

**Comment by jakubch1** [comment]
You can check how often people are using option to collect for test projects here (this is for coverlet): https://github.com/search?p=4&q=IncludeTestAssembly&type=Code

Microsoft.CodeCoverage is able to collect code coverage for both .NET and C++. In C++ world you often link statically to be tested code into test project so it is required in this case to enable coverage for test projects.

For easier usage we could add `ExcludeTestAssembly` flag on main level in configuration. Then usage could be for example:
```
dotnet test --collect "Code Coverage;ExcludeTestAssembly=true"
```
@StingyJack @bramborman @MarcoRossignoli  WDYT?

**Comment by StingyJack** [comment]
I think that trying to determine how well the code that is executed in production is covered by unit tests is valuable. 
[Unit Tests]
    [Prod Code] <-- we care about the coverage of this

I think that trying to calculate how well those unit tests are covered by an additional layer of unit tests is a waste of time. 
[Unit Tests] <-- nobody is writing this, because its turtles all the way down (or up)
  [Unit Tests] <-- who cares about the coverage of this
    [Prod Code] 

Anyone using coverage calculation in the latter example is probably doing so by mistake or has been misinformed.

**Comment by MarcoRossignoli** [comment]
> dotnet test --collect "Code Coverage;ExcludeTestAssembly=true"

I think that this is can be a good option, anyway I find also good reason sometime to have coverage of the test module, you can catch errors(mistakes in test filtering) that are not possible to catch simply looking to the production code.

We had similar discussion on coverlet side https://github.com/coverlet-coverage/coverlet/issues/618

**Comment by bramborman** [comment]
@StingyJack I don't think (hope 😄) anybody writes tests that would test other tests. That would make sense probably only if you have a huge custom test framework. From what I've seen in my results, the test projects covered themselves.

Makes sense that by getting code coverage of the tests, you see if you have some parts of tests that don't get executed. But I'd call this maybe an advanced thing. And you definitely don't want to raise the overall percentage of covered code by the coverage of the tests themselves. Personally I'd like to have two values - one for production code coverage, one for the test code coverage, should I ever need to have the value for tests. But this is of course another story.

I can see that it would be a breaking change to change this behavior, wonder for how many people/projects, but still a breaking change so probably the best that can be done is to provide a simple way to control this behavior, as suggested before.

**Comment by jakubch1** [comment]
We are considering adding new flag to easily exclude test projects so reopening this issue.
